### PR TITLE
Add `filter_map_parallel`

### PR DESCRIPTION
**Pull request by ripytide**
I've re-based the #608 pr since it was ready for merging and don't want it to be left behind.

### CODE REVIEW

File: src/filter/mod.rs [hunk 1]
```diff
@@ -578,7 +663,7 @@ mod tests {
             #[test]
             fn $test_name() {
                 // I think the interesting edge cases here are determined entirely
-                // by the relative sizes of the kernel and the image side length, so
+                // by the relative sizes of the kernel and the image side length, soz
```
Comment (theotherphil) [inline src/filter/mod.rs:666]:
```suggestion
                // by the relative sizes of the kernel and the image side length, so
```

**Comment by sunsided** [comment]
Thanks!

**Comment by theotherphil** [comment]
Thanks @sunsided and @ripytide
